src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -16985,6 +16985,17 @@ void Unit::SetDisplayId(uint32 modelId)
     if (CreatureModelInfo const* minfo = sObjectMgr->GetCreatureModelInfo(modelId))
         SetByteValue(UNIT_FIELD_BYTES_0, 2, minfo->gender);
 
+    // Set Race by modelId for non-players
+    if (GetTypeId() != TYPEID_PLAYER)
+    {
+        if (CreatureDisplayInfoEntry const* display = sCreatureDisplayInfoStore.LookupEntry(modelId))
+        {
+            if (CreatureDisplayInfoExtraEntry const* displayExtra = sCreatureDisplayInfoExtraStore.LookupEntry(display->ExtendedDisplayInfoID))
+            {
+                setRace(displayExtra->DisplayRaceID);
+            }
+        }
+    }
```

Comment:
should probably be done in a separate function that you can call in Creature::InitEntry as this technically has nothing to do with SetDisplayId. And since it is only for creatures we don't need it in unit.

Something like Creature::SetRaceByModelId(uint32 modelId)

Reply:
I will move it to creature.